### PR TITLE
ENH: add isinf, isnan, fmin, fmax loops for datetime64, timedelta64

### LINKED ISSUE #14831: `np.isnan` raises an error for datetime (NaT)
### Reproducing code example:

```python
import numpy as np
a = np.array(['1999-12-15', 'NaT'], dtype='M8[ns]')

np.isnan(a)
np.nanmin(a)
np.nanmax(a)
```

### Error message:

Each of those raises an: 
TypeError: ufunc 'isnan' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''

### Background

Numpy 1.18 returns `NaT` for `np.min(a)` (it did skip `NaT` before 1.18; #14717). This breaks xarray (pydata/xarray#3409) and pandas (pandas-dev/pandas#29058) - therefore it would be good to make `np.isnan` work with `datetime`.

Also: #9009, #1307, #5222

### Numpy/Python version information:
1.18.0.dev0+3b5db53 3.7.3 | packaged by conda-forge | (default, Jul  1 2019, 21:52:21)
[GCC 7.3.0]

### PR DESCRIPTION
**Pull request by mattip**
Fixes gh-14831.

Changes behavior. Previously `isnan`, `isinf` were unavailable for `datetime64` dtypes. Now `assert_equal` behaves slightly differently. Previously `assert_equal(a, b)` would raise if `a` and `b` contained `NaT` in the same position. Now `NaT` is equivalent to `float('nan')` and the call passes. I think this is consistent with the function docstring:

> This function handles NaN comparisons as if NaN was a "normal" number.
    That is, no assertion is raised if both objects have NaNs in the same
    positions.  This is in contrast to the IEEE standard on NaNs, which says
    that NaN compared to anything must return False.

### CODE REVIEW

**Comment by mattip** [comment]
In light of the failure of gh-14800, we should make sure this PR does not cause problems on pandas.

**Comment by mattip** [comment]
Tested this against pandas 0.25.3. Some tests will need adjustment since now they do not raise, but it did not seem to cause new failures:
```
FAILED pandas/tests/indexes/test_numpy_compat.py::test_numpy_ufuncs_other[DatetimeIndex-isinf] - Failed: DID NOT RAISE <class 'Except...
FAILED pandas/tests/indexes/test_numpy_compat.py::test_numpy_ufuncs_other[DatetimeIndex-isnan] - Failed: DID NOT RAISE <class 'Except...
FAILED pandas/tests/indexes/test_numpy_compat.py::test_numpy_ufuncs_other[TimedeltaIndex-isinf] - Failed: DID NOT RAISE <class 'Excep...
FAILED pandas/tests/indexes/test_numpy_compat.py::test_numpy_ufuncs_other[TimedeltaIndex-isnan] - Failed: DID NOT RAISE <class 'Excep...
```

File: numpy/core/src/umath/loops.c.src [hunk 1]
```diff
@@ -1245,6 +1245,22 @@ NPY_NO_EXPORT void
     }
 }
 
+
+NPY_NO_EXPORT void
+@TYPE@_isnan(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUSED(func))
```
Comment (eric-wieser) [inline numpy/core/src/umath/loops.c.src]: Tempting to just use `#define @TYPE@_isnan @TYPE@_isnat`

File: numpy/core/tests/test_datetime.py [hunk 2]
```diff
@@ -2232,26 +2232,29 @@ def test_isnat_error(self):
...
-            arr = np.array([123, -321, "NaT"], dtype='>timedelta64[%s]' % unit)
-            assert_equal(np.isfinite(arr), res)
+    @pytest.mark.parametrize('unit', ['Y', 'M', 'W', 'D', 'h', 'm', 's', 'ms',
+                                      'us', 'ns', 'ps', 'fs', 'as'])
+    def test_isxx_units(self, unit):
+        '''check isfinite, isinf, isnan for all units of <M, >M, <m, >m dtypes
+        '''
+        arr_val = [123, -321, "NaT"]
+        dstr = ['<datetime64[%s]', '>datetime64[%s]',
+                '<timedelta64[%s]', '>timedelta64[%s]']
```
Comment (eric-wieser) [inline numpy/core/tests/test_datetime.py]: Any reason not to parametrize on these too?

File: numpy/core/tests/test_datetime.py [hunk 3]
```diff
@@ -2232,26 +2232,29 @@ def test_isnat_error(self):
...
-            assert_equal(np.isfinite(arr), res)
-            arr = np.array([123, -321, "NaT"], dtype='>datetime64[%s]' % unit)
-            assert_equal(np.isfinite(arr), res)
-            arr = np.array([123, -321, "NaT"], dtype='<timedelta64[%s]' % unit)
-            assert_equal(np.isfinite(arr), res)
-            arr = np.array([123, -321, "NaT"], dtype='>timedelta64[%s]' % unit)
-            assert_equal(np.isfinite(arr), res)
+    @pytest.mark.parametrize('unit', ['Y', 'M', 'W', 'D', 'h', 'm', 's', 'ms',
+                                      'us', 'ns', 'ps', 'fs', 'as'])
+    def test_isxx_units(self, unit):
```
Comment (eric-wieser) [inline numpy/core/tests/test_datetime.py]: for the sake of searchability, perhaps:
```suggestion
    def test_isfinite_isinf_isnan_units(self, unit):
```

File: numpy/testing/_private/utils.py [hunk 4]
```diff
@@ -386,7 +386,7 @@ def assert_equal(actual, desired, err_msg='', verbose=True):
             if not signbit(desired) == signbit(actual):
                 raise AssertionError(msg)
 
-    except (TypeError, ValueError, NotImplementedError):
+    except (TypeError, ValueError, NotImplementedError, DeprecationWarning):
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: This seems risky, and at least deserving of a source comment.

Comment (mattip) [reply numpy/testing/_private/utils.py]: Added an explanation and a version changed

**Comment by mathause** [comment]
Thanks! Note that `nanmax` and `nanmin` don't return the correct result, even after this change (due to #8975 ?) : 

``` python
import numpy as np
a = np.array(['1999-12-15', 'NaT'], dtype='M8[ns]')
np.nanmax(a) 
np.nanmin(a)
```

Both return `numpy.datetime64('NaT')` Does this belong here or in a separate issue?

**Comment by mattip** [comment]
the issue is that fmin, fmax are mapped to minimum, maximum when they should have unique loops to scan for NaT. I can add it to this PR.

File: numpy/testing/_private/utils.py [hunk 5]
```diff
@@ -386,7 +386,12 @@ def assert_equal(actual, desired, err_msg='', verbose=True):
...
+        # version 1.18:
+        # datetime64/timedelta64 used to raise since they had no isnan
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: To be clear - `gisnan` would raise, or `signbit` would raise?

File: numpy/testing/_private/utils.py (hunk 4 above)
Comment (eric-wieser) [reply numpy/testing/_private/utils.py]: It's tempting to add an entirely separate `except DeprecationWarning` with this comment

File: numpy/testing/_private/utils.py (hunk 5 above)
Comment (mattip) [reply numpy/testing/_private/utils.py]: `gisnan` used to raise and no longer does. `signbit` still raises. I will try to clarify

File: numpy/testing/_private/utils.py (hunk 4 above)
Comment (eric-wieser) [reply numpy/testing/_private/utils.py]: Thinking some more, we should not be catching warnings as exceptions at all - shouldn't we be suppressing them even if they're not exceptions (or not suppressing them at all)?

File: doc/release/upcoming_changes/14841.compatibility.rst [hunk 6]
```diff
@@ -1,6 +1,11 @@
...
+Add more ufunc loops for ``datetime64``, ``timedelta64``
+--------------------------------------------------------
+``np.datetime('NaT')`` should behave more like ``float('Nan')``. Add needed
+infrastructure so ``np.isinf(a)`` and ``np.isnan(a)`` will run on
+``datetime64`` and ``timedelta64`` dtypes. Also added specific loops for
+``fmin`` and ``fmax`` that mask ``NaT``. This may require adjustment to user-
+facing code. Specifically, code that either disallowed the calls to `isinf`` or
+``isnan`` or checked that they raised an exception will require adaptation, and
+code that mistakenly called ``fmax`` and ``fmin`` instead of ``maximum`` or
+``minimum`` respectively will requre adjustment. This also affects ``nanmax``
```
Comment (mathause) [inline doc/release/upcoming_changes/14841.compatibility.rst]: I think nanmax raised before 1.18. (Or maybe I missunderstand this comment).

File: numpy/testing/_private/utils.py [hunk 7]
```diff
@@ -388,6 +388,12 @@ def assert_equal(actual, desired, err_msg='', verbose=True):
 
     except (TypeError, ValueError, NotImplementedError):
         pass
+    except DeprecationWarning:
+        # version 1.18
+        # gisnan used to raise for datetime64/timedelta64 (dt64) since they had
+        # no isnan implementation. That call no longer raises. However, a
+        # DeprecationWarning is raised when comparing dt64 to 0.
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: [this comment](https://github.com/numpy/numpy/pull/14841#discussion_r344258043) still applies - this will still emit noisy warnings if I don't have warnings set to raise - we should either:
* Suppress warnings, rather than catching them
* Locally promote warnings to raise
* Special case datetime types to avoid this path
* Call `cancast` somehow to avoid this path

File: numpy/testing/_private/utils.py [hunk 8]
```diff
@@ -388,6 +388,12 @@ def assert_equal(actual, desired, err_msg='', verbose=True):
...
+        pass
 
     try:
         isdesnat = isnat(desired)
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: The assertion error on line 402/408 below is now unreachable - `assert_equal(datetime64('nat'), timedelta64('nat'))` used to assert and no longer does.

File: doc/release/upcoming_changes/14841.compatibility.rst (hunk 6 above)
Comment (mattip) [reply doc/release/upcoming_changes/14841.compatibility.rst]: fixing

File: numpy/testing/_private/utils.py (hunk 7 above)
Comment (mattip) [reply numpy/testing/_private/utils.py]: special casing seems like the easiest

File: numpy/testing/_private/utils.py (hunk 8 above)
Comment (mattip) [reply numpy/testing/_private/utils.py]: The logic on lines 379-382 now fails, since `gisnan` returns True for Nat as well as Nan. Adding a test and rearranging the logic to test for NaT before Nan.

File: numpy/testing/_private/utils.py [hunk 9]
```diff
@@ -404,6 +389,31 @@ def assert_equal(actual, desired, err_msg='', verbose=True):
...
+        isdesnan = gisnan(desired)
+        isactnan = gisnan(actual)
+        if isdesnan and isactnan:
+            return  # both nan, so equal
+
+        # handle signed zero specially for floats
+        array_actual = array(actual)
+        array_desired = array(desired)
+        if (array_actual.dtype.char in 'Mm' or
+                array_desired.dtype.char in 'Mm'):
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: You only want `M` here, not `m`

File: numpy/testing/tests/test_utils.py [hunk 10]
```diff
@@ -300,9 +300,9 @@ def test_nat_items(self):
             self._test_not_equal([a], b)
 
         for a, b in itertools.product(tds, dts):
-            self._test_not_equal(a, b)
+            self._assert_func(a, b)
             self._test_not_equal(a, [b])
-            self._test_not_equal([a], [b])
+            self._assert_func([a], [b])
```
Comment (eric-wieser) [inline numpy/testing/tests/test_utils.py]: Why this change? Seems weird when the rest of the tests are all `_test_not_equal`

File: numpy/testing/_private/utils.py (hunk 9 above)
Comment (mattip) [reply numpy/testing/_private/utils.py]: It doesn't really matter, the test just below this is only for things that will have a signbit on a 0 value, so timedelta64 might as well never go to the next clause

Comment (eric-wieser) [reply numpy/testing/_private/utils.py]: Well at that point you may as well add integers here too - at any rate, the comment is saying "this won't work", not "this is a shortcut to save time".

Comment (mattip) [reply numpy/testing/_private/utils.py]: ok, fixing

Comment (mattip) [reply numpy/testing/_private/utils.py]: actually, `np.timedelta64(7, 's') == 0` emits a DeprecationWarning, so we need both 'Mm', and the comment is wrong

File: numpy/testing/tests/test_utils.py (hunk 10 above)
Comment (mattip) [reply numpy/testing/tests/test_utils.py]: reverted

**Comment by mathause** [comment]
Do you plan to get this into 1.18? I guess this could help external libraries so that they don't have to check for 3 cases 

* before 1.18: `min -> value`, `nanmin -> error`
* after 1.18: `min -> NaT`, `nanmin -> error`
* after this gets merged: `min -> NaT`, `nanmin -> value`

File: numpy/testing/_private/utils.py (hunk 9 above)
Comment (eric-wieser) [reply numpy/testing/_private/utils.py]: Seems weird that `td[s] == 0` is deprecated yet `td[s] + 0` is not - perhaps worth opening a new issue about.

File: numpy/testing/_private/utils.py [hunk 11]
```diff
@@ -404,6 +389,31 @@ def assert_equal(actual, desired, err_msg='', verbose=True):
...
+    # Inf/nan/negative zero handling
+    try:
+        isdesnan = gisnan(desired)
+        isactnan = gisnan(actual)
+        if isdesnan and isactnan:
+            return  # both nan, so equal
+
+        # handle signed zero specially for floats
+        array_actual = array(actual)
+        array_desired = array(desired)
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: nit: maybe `asarray`?

**Review by eric-wieser — APPROVED** [review]

File: doc/release/upcoming_changes/14841.compatibility.rst [hunk 12]
```diff
@@ -0,0 +1,11 @@
+Add more ufunc loops for ``datetime64``, ``timedelta64``
+--------------------------------------------------------
+``np.datetime('NaT')`` should behave more like ``float('Nan')``. Add needed
+infrastructure so ``np.isinf(a)`` and ``np.isnan(a)`` will run on
+``datetime64`` and ``timedelta64`` dtypes. Also added specific loops for
+``fmin`` and ``fmax`` that mask ``NaT``. This may require adjustment to user-
```
Comment (eric-wieser) [inline doc/release/upcoming_changes/14841.compatibility.rst]:
```suggestion
`fmin` and `fmax` that mask ``NaT``. This may require adjustment to user-
```
may as well link these, right?

File: numpy/testing/_private/utils.py (hunk 11 above)
Comment (mattip) [reply numpy/testing/_private/utils.py]: There is a check above for `isinstance(actual, ndarray) or isinstance(desired, ndarray): return assert_array_equal(actual, desired, ...`. I copied the pattern from (after the changes) line 380 in the clause above this one, should I change them both?

File: doc/release/upcoming_changes/14841.compatibility.rst (hunk 12 above)
Comment (mattip) [reply doc/release/upcoming_changes/14841.compatibility.rst]: ok

File: doc/release/upcoming_changes/14841.compatibility.rst [hunk 13]
```diff
@@ -0,0 +1,11 @@
+Add more ufunc loops for ``datetime64``, ``timedelta64``
+--------------------------------------------------------
+``np.datetime('NaT')`` should behave more like ``float('Nan')``. Add needed
+infrastructure so ``np.isinf(a)`` and ``np.isnan(a)`` will run on
+``datetime64`` and ``timedelta64`` dtypes. Also added specific loops for
+`numpy.fmin` and `numpy.fmax` that mask ``NaT``. This may require adjustment to user-
+facing code. Specifically, code that either disallowed the calls to
+`numpy.isinf`` or `numpy.isnan`` or checked that they raised an exception will
```
Comment (eric-wieser) [inline doc/release/upcoming_changes/14841.compatibility.rst]:
```suggestion
`numpy.isinf` or `numpy.isnan` or checked that they raised an exception will
```

File: numpy/testing/_private/utils.py (hunk 11 above)
Comment (eric-wieser) [reply numpy/testing/_private/utils.py]: Fair enough, I suppose that means there's no point using asarray.

**Review by WarrenWeckesser — CHANGES_REQUESTED** [review]
Just one change requested.  Otherwise LGTM.

File: numpy/testing/_private/utils.py [hunk 14]
```diff
@@ -404,6 +389,31 @@ def assert_equal(actual, desired, err_msg='', verbose=True):
...
+        if (array_actual.dtype.char in 'Mm' or
+                array_desired.dtype.char in 'Mm'):
+            # version 1.18
+            # until this version, gisnan failed for datetime64 and timedelta64.
+            # Now it succeeds but comparison to 0 emits a DeprecationWarning
+            # Avoid that by skipping the next check
+            raise NotImplementedError('cannot compare datetime64 to 0')
```
[the remark this answers is not in the document]
Comment (mattip) [reply numpy/testing/_private/utils.py]: you prefer a bare `raise NotImplementedError()` to the current exception which is explicitly caught on line 414?

[the remark this answers is not in the document]
Comment (WarrenWeckesser) [reply numpy/testing/_private/utils.py]: The comment should be updated, too.  It is not just comparison to 0 that results in the `DeprecationWarning`.

[the remark this answers is not in the document]
Comment (WarrenWeckesser) [reply numpy/testing/_private/utils.py]: Since the exception will be caught a few lines down, is there are need for the message?  A message is fine if you prefer having one, but it should be accurate.  The current message (about comparing to 0) is misleading.

File: numpy/testing/_private/utils.py [hunk 15]
```diff
@@ -404,6 +389,31 @@ def assert_equal(actual, desired, err_msg='', verbose=True):
...
+            return  # both nan, so equal
+
+        # handle signed zero specially for floats
+        array_actual = array(actual)
+        array_desired = array(desired)
+        if (array_actual.dtype.char in 'Mm' or
+                array_desired.dtype.char in 'Mm'):
+            # version 1.18
+            # until this version, gisnan failed for datetime64 and timedelta64.
+            # Now it succeeds but comparison to 0 emits a DeprecationWarning
```
Comment (WarrenWeckesser) [inline numpy/testing/_private/utils.py]:
```suggestion
            # Now it succeeds but comparison to scalar with a different type
            # emits a DeprecationWarning.
```

File: numpy/testing/_private/utils.py (hunk 14 above)
Comment (WarrenWeckesser) [inline numpy/testing/_private/utils.py]:
```suggestion
            raise NotImplementedError('cannot compare to a scalar '
                                      'with a different type')
```

**Review by WarrenWeckesser — CHANGES_REQUESTED** [review]
I suggested a some tweaks to fix the slightly misleading comment and exception message.

**Review by WarrenWeckesser — APPROVED** [review]
